lib/eventHubReceiver.ts

```diff
@@ -0,0 +1,382 @@
+    }
+
+    if (maxWaitTimeInSeconds === null || maxWaitTimeInSeconds === undefined) {
+      maxWaitTimeInSeconds = Constants.defaultOperationTimeoutInSeconds;
+    }
+
+    if (!this._session && !this._receiver) {
+      throw Errors.translate({ condition: Errors.ConditionStatusMapper[404], description: "The messaging entity underlying amqp receiver could not be found." });
+    }
+
+    try {
+      let eventDatas: EventData[] = [];
+      let count = 0;
+      let timeOver = false;
+      return new Promise<EventData[]>((resolve, reject) => {
+        let onReceiveMessage: Func<EventData, void>;
+        let waitTimer: NodeJS.Timer;
+        let actionAfterWaitTimeout: Func<void, void>;
+        // Final action to be performed after maxMessageCount is reached or the maxWaitTime is over.
+        const finalAction = (timeOver: boolean, data?: EventData) => {
+          // Remove the listener to avoid receiving duplicate messages.
+          this.removeListener(Constants.message, onReceiveMessage);
+          if (!data) {
+            data = eventDatas.length ? eventDatas[eventDatas.length - 1] : undefined;
+          }
+          if (timeOver) {
+            clearTimeout(waitTimer);
+          }
+          if (this.receiverRuntimeMetricEnabled && data) {
+            this.runtimeInfo.lastSequenceNumber = data.lastSequenceNumber;
+            this.runtimeInfo.lastEnqueuedTimeUtc = data.lastEnqueuedTime;
+            this.runtimeInfo.lastEnqueuedOffset = data.lastEnqueuedOffset;
+            this.runtimeInfo.retrievalTime = data.retrievalTime;
+          }
+          resolve(eventDatas);
+        };
+
+        // Action to be performed after the max wait time is over.
+        actionAfterWaitTimeout = () => {
+          timeOver = true;
+          finalAction(timeOver);
+        };
+
+        // Action to be performed on the "message" event.
+        onReceiveMessage = (data: EventData) => {
+          if (!timeOver && count <= maxMessageCount) {
+            count++;
+            // console.log(`${new Date().toString()} - ${count}`);
+            eventDatas.push(data);
+          }
+          if (count === maxMessageCount) {
+            finalAction(timeOver, data);
+          }
+        };
+        waitTimer = setTimeout(actionAfterWaitTimeout, (maxWaitTimeInSeconds as number) * 1000);
+        this.on(Constants.message, onReceiveMessage);
+      });
+    } catch (err) {
+      return Promise.reject(err);
+    }
+  }
+
+  /**
+   * Closes the underlying AMQP receiver.
+   */
+  async close(): Promise<void> {
+    if (this._receiver) {
+      try {
+        // TODO: should I call _receiver.detach() or _receiver.close()?
+        // should I also call this._session.close() after closing the reciver
+        // or can I directly close the session which will take care of closing the receiver as well.
+        await rheaPromise.closeReceiver(this._receiver);
```

Comment:
@grs - It would be awesome if you could review the last two commits in the PR that contain the source code and tests. I will be calling you out at specific sections for questions specific to rhea. Please let me know your thoughts on the questions asked in the code comments.